# Movies Tonight Part 3: SQL DML

In [2]:
%load_ext sql
%sql sqlite:///Movies.db

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


'Connected: @Movies.db'

1. **Load the `DATASET.csv` file into the `Movies.db` database.**  
    - Import the CSV file to create a new table called `DATASET`. Follow the example [here](http://www.sqlitetutorial.net/sqlite-import-csv/).
    - Use a [`PRAGMA table-info` pseudo-query](https://www.sqlite.org/pragma.html#pragma_table_info) to inspect the design of the new table. [How are we going to handle dates](https://www.sqlite.org/datatype3.html)?

In [ ]:
YOUR CODE HERE

2. **Populate the tables with data.**  

    - Use `INSERT ... SELECT ...` queries to extract data from `DATASET` for each table of your other tables. 
    - Start with the strong entities before moving on to the weak ones. Be sure to always select distinct     values. 
    - Foreign Keys get special treatment. You will have to use JOINs to pick up the PKs from the other side of the relationships. (As in Part 1, use the fact that movie names, theater names, and people names are always unique in our source data. That way you can use then for the JOIN columns. 
    - If you have to start over, then rerun your DDL queries from Part 2. 

In [4]:
%%sql
INSERT INTO MOVIES (Mtitle, Rating)
SELECT DISTINCT Mtitle, Rating FROM DATASET;

 * sqlite:///Movies.db
23 rows affected.


[]

In [6]:
%%sql
INSERT INTO THEATRE (Tname, Phone, Location)
SELECT DISTINCT Tname, Phone, Location FROM DATASET;

 * sqlite:///Movies.db
9 rows affected.


[]

In [11]:
%%sql
INSERT INTO ARTIST (Name)
SELECT DISTINCT CName FROM DATASET;

 * sqlite:///Movies.db
153 rows affected.


[]

In [15]:
%%sql
INSERT INTO CREDIT (MID, AID, Ccode)
SELECT DISTINCT MID, AID, Ccode 
FROM DATASET 
    JOIN MOVIES ON (MOVIES.MTitle = DATASET.Mtitle)
    JOIN ARTIST ON (ARTIST.Name = DATASET.CName);


 * sqlite:///Movies.db
162 rows affected.


[]

In [20]:
%%sql
INSERT INTO SHOW (MID, TID, STime)
SELECT DISTINCT MID, TID, ShowTime
FROM DATASET 
    JOIN MOVIES ON (DATASET.MTitle = MOVIES.Mtitle)
    JOIN THEATRE ON (DATASET.TName = THEATRE.Tname);

 * sqlite:///Movies.db
131 rows affected.


[]

3. **Check your work.**  
There are 9 theaters, 23 movies, 152 artists, 131 shows, and 161 credits.

4. **Now check the source data.**  
There are actually 13 duplicate rows in the original source data! Can you spot them? Write a SQL query that counts how many times each (MTitle, TName, Showtime, CCode, and CName) appears. Which ones appear more than once?

In [41]:
%%sql
SELECT MTitle, TName, ShowTime, CCode, CName, COUNT(*)
FROM DATASET
GROUP BY MTitle, TName, ShowTime, CCode, CName 
HAVING COUNT(*) > 1



 * sqlite:///Movies.db
Done.


[('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Greg Kinnear', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Hector Elizondo', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Jon Seda', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Laurie Metcalf', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Maria Pitillo', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Roscoe Lee Browne', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Tim Conway', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'D', 'Garry Marshall', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'Jon Lovitz', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'Mekhi Phifer', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'Tia Carrera', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'and Louise Fletcher', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'D', 'Hart Bochner', 2)]

5. **Drop the `DATASET` table.**  
We don't need it anymore. 

In [42]:
%sql DROP TABLE DATASET;

 * sqlite:///Movies.db
Done.


[]

6. **Save, commit, and push your work to GitHub.**  
Use the commit summary comment "Completed Part 3".